File: time-travel/ReportProjection/ReportReadModel.cs
using System;
using System.Collections.Generic;
using System.Text.Json.Serialization;

using Region = string;
using Category = string;
using ReportDate = string;

namespace ReportProjection
{
    // Root type for deserialization
    public record ReportReadModel
    {
        [JsonPropertyName("checkpoint")]
        public long Checkpoint { get; set; }

        [JsonPropertyName("salesReports")]
        public Dictionary<ReportDate, SalesReport> SalesReports { get; init; } = new();
    }
    
    public record SalesReport
    {
        [JsonPropertyName("categorySalesReports")]
        public Dictionary<Category, CategorySalesReport> CategorySalesReports { get; init; } = new();
    }
    
    public record CategorySalesReport
    {
        [JsonPropertyName("regionSalesReports")]
        public Dictionary<Region, RegionSalesReport> RegionSalesReports { get; init; } = new();
   }
    
    public record RegionSalesReport
    {
        [JsonPropertyName("dailySales")]
        public decimal DailySales { get; init; }

        [JsonPropertyName("targetSales")]
        public decimal TargetSales { get; init; }

        [JsonPropertyName("totalMonthlySales")]
        public decimal TotalMonthlySales { get; init; }

        [JsonPropertyName("targetHitRate")]
        public decimal TargetHitRate { get; init; }
    }
}

File: time-travel/DemoWeb/client/app/ReportReadModel.ts
export type Region = string;
export type Category = string;
export type ReportDate = string;

// Root type for deserialization
export interface ReportReadModel {
  checkpoint: number;
  salesReports: Record<ReportDate, SalesReport>;
}

export interface SalesReport {
  categories: Record<Category, CategorySalesReport>;
}

export interface CategorySalesReport {
  regions: Record<Region, RegionSalesReport>;
}

export interface RegionSalesReport {
  dailySales: number;
  monthEndSalesTarget: number;
  totalMonthlySales: number;
  targetHitRate: number;
}
